### PR TITLE
feat: support calling V1 TES when useTesV1 Nucleus config is true

### PR DESCRIPTION
**Pull request by MikeDombo**
**Issue #, if available:**

**Description of changes:**
This change adds a new configuration in the Nucleus component config called "useTesV1". When true, the TES URL path is changed to be the path which the Greengrass V1 cloud offers. The customer can use this in conjunction with setting the `iotCredEndpoint` to the Greengrass ats data endpoint in order to retrieve credentials from the V1 TES. If a customer chooses to do this, they must have a Greengrass group setup with the role associated, a core created with the Greengrass V2 thing's certificate ARN, and must have a deployment created for the group.

**Why is this change necessary:**
This is necessary for customers who want to use Greengrass in a VPC with VPCE as IoT Credential Provider doesn't yet support VPCE.

**How was this change tested:**
Added basic E2E test code (not committed) which created a Greengrass group with role and then verified that we got the credentials from it.

**Any additional information or context required to review the change:**

**Checklist:**
 - [ ] Updated the README if applicable
 - [x] Updated or added new unit tests
 - [ ] Updated or added new integration tests
 - [ ] Updated or added new end-to-end tests
 - [ ] If your code makes a remote network call, it was tested with a proxy
 - [x] You confirm that the change is backwards compatible

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/aws/greengrass/tes/TokenExchangeService.java [hunk 1]
```diff
@@ -37,6 +38,9 @@
     private static final String TES_CONFIG_ERROR_STR = "%s parameter is either empty or not configured for TES";
     // randomly choose a port
     private static final int DEFAULT_PORT = 0;
+    // Use a time in the far future so that when we use withNewerValue, the configuration change will always happen,
+    // no matter what the current system clock is (as long as the clock isn't ahead of 3000)
+    private static final long FUTURE = Instant.parse("3000-01-01T00:00:00Z").toEpochMilli();
```
Comment (jbutler) [inline src/main/java/com/aws/greengrass/tes/TokenExchangeService.java:43]: Can you break this into a separate commit? I'd prefer having a commit message that adds a little more context

File: src/main/java/com/aws/greengrass/tes/CredentialRequestHandler.java [hunk 2]
```diff
@@ -115,16 +126,31 @@ public CredentialRequestHandler(final IotCloudHelper cloudHelper, final IotConne
...
         deviceConfiguration.getRootCAFilePath().subscribe((why, newv) -> clearCache());
         deviceConfiguration.getPrivateKeyFilePath().subscribe((why, newv) -> clearCache());
+        Topic tesV1Topic = deviceConfiguration.getNucleusConfig().lookup("useTesV1");
+        tesV1Topic.subscribe((why, newv) -> {
+            if (tesV1Topic.getOnce() == null) {
+                setIotCredentialsPath(Coerce.toString(deviceConfiguration.getIotRoleAlias()));
+            } else {
+                if (Coerce.toBoolean(tesV1Topic)) {
+                    this.iotCredentialsPath = "/greengrass/assumeRoleForGroup";
+                }
```
Comment (jbutler) [inline src/main/java/com/aws/greengrass/tes/CredentialRequestHandler.java:151]: If the customer sets `useTesV1` to true and then to false, won't this fail to set the credential path back to V2 TES?

File: src/main/java/com/aws/greengrass/tes/TokenExchangeService.java (hunk 1 above)
Comment (MikeDombo) [reply src/main/java/com/aws/greengrass/tes/TokenExchangeService.java:43]: Different commit, same PR, done.

File: src/main/java/com/aws/greengrass/tes/CredentialRequestHandler.java (hunk 2 above)
Comment (MikeDombo) [reply src/main/java/com/aws/greengrass/tes/CredentialRequestHandler.java]: Done.

File: src/main/java/com/aws/greengrass/tes/CredentialRequestHandler.java [hunk 3]
```diff
@@ -115,16 +126,29 @@ public CredentialRequestHandler(final IotCloudHelper cloudHelper, final IotConne
 
         deviceConfiguration.getIotRoleAlias().subscribe((why, newv) -> {
             String iotRoleAlias = Coerce.toString(newv);
-            clearCache();
-            setIotCredentialsPath(iotRoleAlias);
+            if (Utils.isNotEmpty(iotRoleAlias)) {
+                clearCache();
+                setIotCredentialsPath(iotRoleAlias);
+            }
```
Comment (jbutler) [inline src/main/java/com/aws/greengrass/tes/CredentialRequestHandler.java:132]: I don't think this is quite right - `useTesV1` needs to be definitive otherwise you get into situations where the order in which config is updated or read changes the behavior, right? Role alias should only be processed if `useTesV1` is absent or false.

Example:
* Customer configures `useTesV1 == true` -> TES switches to V1
* Customer updates `iotRoleAlias` -> TES switches back to V2
* Customer restarts Nucleus -> TES processes `useTesV1` last and uses V1

Comment (jbutler) [reply src/main/java/com/aws/greengrass/tes/CredentialRequestHandler.java:132]: It may be a stop gap solution, but once we ship it, we can't remove it.

I can see this being an issue after a customer migrates from V1 back to V2. From an ops perspective, we need to be able to know what GG is doing based on just the config. Should be simple to fix, no?

Comment (MikeDombo) [reply src/main/java/com/aws/greengrass/tes/CredentialRequestHandler.java]: Updated

**Review by jbutler — APPROVED** [review]

File: src/test/java/com/aws/greengrass/tes/CredentialRequestHandlerTest.java [hunk 4]
```diff
@@ -78,8 +78,8 @@
     // Set in the far future so it doesn't expire
     private static final String EXPIRATION = "2030-08-19T07:35:15Z";
     private static final String RESPONSE_STR =
-            "{\"credentials\":" + "{\"accessKeyId\":\"" + ACCESS_KEY_ID + "\"," + "\"secretAccessKey\":\""
-                    + SECRET_ACCESS_KEY + "\"," + "\"sessionToken\":\"" + SESSION_TOKEN + "\"," + "\"expiration\":\""
+            "{\"credentials\":" + "{\"AccessKeyId\":\"" + ACCESS_KEY_ID + "\"," + "\"secretAccessKey\":\""
+                    + SECRET_ACCESS_KEY + "\"," + "\"sesSionToken\":\"" + SESSION_TOKEN + "\"," + "\"expiration\":\""
```
Comment (shaguptashaikh) [inline src/test/java/com/aws/greengrass/tes/CredentialRequestHandlerTest.java:82]: why?

Comment (MikeDombo) [reply src/test/java/com/aws/greengrass/tes/CredentialRequestHandlerTest.java:82]: Because V1 TES uses different capitalization, I wanted to ensure that we don't care how it is capitalized.

**Review by shaguptashaikh — APPROVED** [review]
LGTM, I'm just a little unsure if 'useTesV1' is very user friendly, don't think customers would be pretty clear on what `TES` means

**Comment by MikeDombo** [comment]
> LGTM, I'm just a little unsure if 'useTesV1' is very user friendly, don't think customers would be pretty clear on what `TES` means

Any alternative suggestions?

**Comment by shaguptashaikh** [comment]
> > LGTM, I'm just a little unsure if 'useTesV1' is very user friendly, don't think customers would be pretty clear on what `TES` means
> 
> Any alternative suggestions?

Nothing great, maybe something like `useV1CredentialEndpoint` or `useV1TokenExchangeService`
